ui-speedspacechart/src/components/helpers/drawElements/declivity.ts

```diff
@@ -0,0 +1,62 @@
+import { clearCanvas, slopesValues, maxPositionValues } from '../../utils';
+import type { Store } from '../../../types/chartTypes';
+import { MARGINS } from '../../const';
+
+const { CURVE_MARGIN_SIDES, MARGIN_TOP, MARGIN_BOTTOM } = MARGINS;
+
+export const drawDeclivity = (
+  ctx: CanvasRenderingContext2D,
+  width: number,
+  height: number,
+  store: Store
+) => {
+  const { slopes, ratioX, leftOffset } = store;
+  console.log('slopes:', { slopes, ratioX, leftOffset, width, height, store }, 'params and store');
+
+  const { maxGradient, slopesRange } = slopesValues(store);
+  const { maxPosition } = maxPositionValues(store);
+
+  if (!slopes || slopes.length === 0) {
+    console.error('Slopes data is missing or empty.');
+    return;
+  }
+
+  clearCanvas(ctx, width, height);
+
+  ctx.save();
+  ctx.translate(leftOffset, 0);
+
+  // Calculate total height available for drawing, excluding the margins
+  const availableHeight = height - MARGIN_TOP - MARGIN_BOTTOM - 30;
+
+  // Calculate the vertical center of the chart
+  const centerY = MARGIN_TOP + 30 + availableHeight / 2;
+
+  console.log('centerY:', centerY, 'maxGradient:', maxGradient);
+
+  ctx.fillStyle = '#CFDDCE';
+
+  try {
+    for (let i = 0; i < slopes.length - 1; i++) {
+      const current = slopes[i];
+      const next = slopes[i + 1];
+
+      const x1 =
+        current.position * ((width - CURVE_MARGIN_SIDES) / maxPosition) * ratioX +
+        CURVE_MARGIN_SIDES / 2;
+      const x2 =
+        next.position * ((width - CURVE_MARGIN_SIDES) / maxPosition) * ratioX +
+        CURVE_MARGIN_SIDES / 2;
+
+      const rectWidth = x2 - x1;
+
+      const rectHeight = (current.gradient / slopesRange) * availableHeight;
+
+      ctx.fillRect(x1, centerY, rectWidth, rectHeight);
+    }
+  } catch (error) {
+    console.error('Error while drawing declivity:', error);
+  }
```

Comment:
Small perf. improvement here : avoid recalculating values that do not change in the loop. For `x1` and `x2`, we can store the second part of calculation in a variable just before the loop. (Not sure if `coef` is the right name for the variable)
```suggestion
  ctx.fillStyle = '#CFDDCE';

  try {
    const coef = ((width - CURVE_MARGIN_SIDES) / maxPosition) * ratioX +
        CURVE_MARGIN_SIDES / 2;
    
    for (let i = 0; i < slopes.length - 1; i++) {
      const current = slopes[i];
      const next = slopes[i + 1];

      const x1 = current.position * coef;
      const x2 = next.position * coef;

      const rectWidth = x2 - x1;
      const rectHeight = (current.gradient / slopesRange) * availableHeight;

      ctx.fillRect(x1, centerY, rectWidth, rectHeight);
    }
  } catch (error) {
    console.error('Error while drawing declivity:', error);
  }
```

Reply:
Good point. I will let curve_margin_sides / 2 in the loop, as it's resulting in incorrect values when applied in the coef.